plugin/pkg/project/validator/admission.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validator
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	gardencore "github.com/gardener/gardener/pkg/apis/core"
+	gutil "github.com/gardener/gardener/pkg/utils/gardener"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apiserver/pkg/admission"
+)
+
+const (
+	// PluginName is the name of this admission plugin.
+	PluginName = "ProjectValidator"
+)
+
+// Register registers a plugin.
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return New()
+	})
+}
+
+type handler struct {
+	*admission.Handler
+}
+
+// New creates a new handler admission plugin.
+func New() (*handler, error) {
+	return &handler{
+		Handler: admission.NewHandler(admission.Create),
+	}, nil
+}
+
+var _ admission.ValidationInterface = &handler{}
+
+func (v *handler) Validate(_ context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	// Ignore all kinds other than Project
+	if a.GetKind().GroupKind() != gardencore.Kind("Project") {
+		return nil
+	}
+
+	// Ignore updates to status or other subresources
+	if a.GetSubresource() != "" {
+		return nil
+	}
+
+	// Convert object to Project
+	project, ok := a.GetObject().(*gardencore.Project)
+	if !ok {
+		return apierrors.NewBadRequest("could not convert object to Project")
+	}
+
+	if project.Spec.Namespace != nil && !strings.HasPrefix(*project.Spec.Namespace, gutil.ProjectNamespacePrefix) {
+		return admission.NewForbidden(a, fmt.Errorf(".spec.namespace must start with %s", gutil.ProjectNamespacePrefix))
+	}
```

Comment:
Is there a reason why we feature a new admission plugin for such small and static validation? Is it possible to perform this validation in `pkg/apis/core/validation/project.go`. There the Project .spec.namespace is already validated in similar way - https://github.com/gardener/gardener/blob/a1c98458113d13b290cfc59404091f3a17fdf70d/pkg/apis/core/validation/project.go#L71-L73. I think the validation there can be improved and we can get rid of this admission plugin. WDYT?

/hold

Reply:
The motivation for the admission plugin was to enable to easily turn on/off this check as it is backwards-incompatible.

Reply:
Then I would suggest to add a TODO with this reason that also states that this check should be moved to `pkg/apis/core/validation/project.go` in the next releases and this admission plugin should be deprecated && removed. WDYT?